Text editor UI tests

phil-davis opened the pull request:
An initial set of UI tests that:create new text files with the text editor, using various filenames and putting bits of text in the files.

To do:

- [x] type in just the file name, leaving the default ".txt" file type
- [x] press the close button to dismiss the editor
- [x] try creating an existing file, check the tooltip comes up
- [x] be able to read back the content of the files to check that it really was saved
- [x] edit an existing text file and make some changes
- [x] navigate into sub-folders and create/edit files there

phil-davis commented:
@PVince81 any chance to add the SauceLabs credentials to Travis for this repo?
These tests run for me locally, and it would be nice to be able to see if they run on Travis/SauceLabs.

phil-davis commented on tests/ui/features/bootstrap/TextEditorContext.php:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	/**
+	 * @When /^I create a text file with the name ((?:'[^']*')|(?:"[^"]*"))( and the default file type|)$/
```
I want ( and the default file type) to be an optional phrase on the end of the step text. So for that I think I have to dive in regex land.
But the name of the file could be in single or double quotes, so I have to look for either a group surrounding by single quotes, or a group surrounded by double quotes.
The regex then captures the whole thing, including the outside pair of quotes.
There are alternative regex that first capture the which quote it is:
```
(["'])
```
and then use that with a back-reference to the captured quote.

But then the quote comes through as the first capture group, and ends up as a parameter to the function!

If if match the first quote with a non-capturing group, then I cannot use a back-reference to it:
```
(?:["'])
```
So what I have done here is the best I have come up with so far.

individual-it replied:
why do you need `and the default file type`? what other file types could a text file have? Do you mean the file extension `.txt`?

phil-davis replied:
The "new text file name" box comes up with "New text file.txt" and "New text file" is highlighted. If you just type straight away, without clearing the field, then it retains the ".txt". So I want to test that behavior, using "the default file type" that is presented by default.
Is it easier to understand if I call it "the default file extension?

individual-it replied:
"the default file extension" or "without changing the default file extension"

phil-davis commented on tests/ui/features/lib/TextEditorPage.php:
```diff
@@ -0,0 +1,138 @@
+
+	/**
+	 * type in the field that matches the given xpath and press enter.
+	 * Note: this depends on methods that might only be in the Selenium implementation
```
I tried to use keyPress, keyUp, keyDown with no joy.
Using postValue works fine, but it is part of the underlying webDriver that does not seem to get exposed by Mink.
Is there some other way to "just type stuff" where the cursor is, without clearing the current contents of the field?

individual-it replied:
yes the keypressing functions are somehow rubbish, had a lot of trouble with them.
you could try to implement raw typing with JS, or somehow manipulating that field with JS.

phil-davis replied:
or I could leave this as it is. and we admit that we depend on Selenium?

individual-it replied:
👍 yes, if you are happy with that solution, just pass in the Session variable from higher up and all is fine. We anyway depend on Selenium

individual-it commented on tests/ui/features/lib/TextEditorPage.php:
```diff
@@ -0,0 +1,138 @@
+* ownCloud
+*
+* @author Phillip Davis
+* @copyright 2017 Phillip Davis info@individual-it.net
```
<phil@jankaritech.com> ?

phil-davis replied:
Copy-paste!

individual-it commented on tests/ui/features/lib/TextEditorPage.php:
```diff
@@ -0,0 +1,138 @@
+	 */
+	public function typeInFieldAndPressEnter($xpath, $text)
+	{
+		$element = $this->getSession()->getDriver()->getWebDriverSession()->element("xpath", $xpath);
```
`getSession()` is Deprecated. Accessing the session from the element is deprecated as of 1.6 and will be impossible in 2.0.
In other places where we need the Session, we do pass it in from the context https://github.com/owncloud/core/blob/master/tests/ui/features/bootstrap/FilesContext.php#L48

phil-davis commented:
Code style fixed up, addressed all comments (I think), squashed the commits.
Can be reviewed again, if you like.
This afternoon, I will try to make the next couple of points in the ToDo. This may as well be a developing PR while waiting to see if SauceLabs CI can be enabled. At that point, I can get the automatic CI to work and then something can be merged.

individual-it commented on tests/ui/features/lib/TextEditorPage.php:
```diff
@@ -0,0 +1,184 @@
+	 */
+	public function typeInFieldAndPressEnter($xpath, $text, Session $session) {
+		$element
+			= $session->getDriver()->getWebDriverSession()->element("xpath", $xpath);
```
why is this new line happening?

phil-davis replied:
Coding standard said the line was to long.
It made me put the "=" on the next line.
Somehow it thinks this is shorter.

individual-it replied:
yes by 1 character :-)

individual-it replied:
that should also work:
```
$session->getDriver()
              ->getWebDriverSession()->element("xpath", $xpath);
```

```
$session->getDriver()->getWebDriverSession()->element(
    "xpath", $xpath
);
```

phil-davis replied:
I think phpcs is a bit stupid about the tabs (i.e. different to how my IDE visually displays them) so on my IDE the tabbed line goes past its 80 column dotted line. But counting tabs as 1, 2 or 4 spaces means the line fits.

PVince81 commented:
kapoot ?

individual-it commented:
@PVince81 not kaput but need Saucelabs credentials in this repo https://github.com/owncloud/files_texteditor/pull/204#issuecomment-324934018

tomneedham commented:
Awesome, thanks for the text editor love  @phil-davis

phil-davis commented:
Happy to have some fun. I am adding some tests that open existing text files and verify that content. And I have code that will navigate into sub-folders and make/check text files in them. I should be able to push all that tonight (my time). Then I think the whole end-to-end testing for this can be merged in this 1 PR.

phil-davis commented:
If somebody can get Saucelabs credentials into the Travis settings for this repo, we can even have the test runs automated.

individual-it commented:
I've added my saucelab credentials to travis. Someone should add owncloud credentials (same as in core) at some point.
@DeepDiver1975  ?

phil-davis commented on tests/ui/features/lib/TextEditorPage.php:
```diff
@@ -0,0 +1,219 @@
+		if ($textLayer === null) {
+			throw new ElementNotFoundException("could not find text layer");
+		}
+		$textLineElements = $textLayer->findAll(
```
@individual-it this returns me an array of elements that are the lines of text, but every element appears twice in the array.
Can you have a look and see what I am doing wrong in the xpath selector or?

(it is really consistent, with the code below to choose the text from every 2nd element, the tests never fail.)

individual-it replied:
I think its because of the xpath `$textFileLineXpath = ".//div[contains(@class,'ace_line')]";`
It selects two divs `div class="ace_line_group"` and `div class="ace_line"`
![image](https://user-images.githubusercontent.com/2425577/29958031-69926912-8f10-11e7-88ef-4c3dfaac5784.png)

what about changing this xpath to: `.//div[@class='ace_line']` then you will have only the inner div

phil-davis replied:
That worked thank you
Now I understand that "contains" is just doing a substr kind of thing. I was hoping it would check through the classes (=words) in the class list of an element and choose if the element "contains" that class.

phil-davis commented on tests/ui/features/textfiles.feature:
```diff
@@ -0,0 +1,96 @@
+		Then the file "New text file.txt" should be listed
+		And I open the file "New text file.txt"
+		Then line 1 of the text is "stuff"
+		And I input "lots of other text so that I can see it working" in the text area
```
TODO:
This is a start towards checking that an edit of an existing file gets saved.
At the moment, the Mink/Selenium code seems to delete the "st" at the front of "stuff" and then type in the given text, so the file ends up with "lots of other text so that I can see it workinguff" in it.

phil-davis commented:
I have done all that I think is relevant for an "enhanced" smoke test.
@SergioBertolinSG @davitol @individual-it please review the whole thing.

individual-it commented on tests/ui/features/bootstrap/TextEditorContext.php:
```diff
@@ -0,0 +1,166 @@
+		// Note: in testing with a 2 second sleep there were still some fails
+		// so it needs quite a long sleep to ensure the latest content has
+		// been saved.
+		sleep(5);
```
could we use the waitTillAjaxStartAndFinish function here?

phil-davis replied:
I will play with it. I think the ACE code just saves at timed intervals when there are changes (i.e. save does not trigger on every input keystroke) - so it is possible to finish typing and nothing will be happening then some (short) time later "save" will wake up.

But I will check.

individual-it commented:
👍

phil-davis commented:
The wait for AJAX things works well before closing the editor. I run the scenario at line 50 lots of times and it always passed, and watching the screen it waits until just after all the "Saving..." messages have stopped flashing up.

So I pushed a commit for that.
